File: 03 - source code/01 - imperative languages/Zig/random_bitstring_and_flexible_password_generator.zig
// random_bitstring_and_flexible_password_generator.zig
//
// 2025-05-14/15/16/22, 2025-06-01, 2025-07-22: fixing error handling when writing to files without a try-catch construct
// 2025-12-19: see below
//
// build on Ubuntu 24 LTS: $ zig build-exe random_bitstring_and_flexible_password_generator.zig -mcpu=native-avx512f -O ReleaseFast
//                           switch -mcpu=native-avx512f is absolutely needed at the test system,
//                           otherwise and even with an "empty" program, running valgrind would crash immediately!
//                           $ valgrind ./random_bitstring_and_flexible_password_generator
//
// run on Ubuntu 24 LTS:   $ ./random_bitstring_and_flexible_password_generator
//
// $ zig version
// 0.15.2
// $


const std = @import("std");

pub fn main() !void {  // ! --> error handling: allow this function to return an error.
                       // So, there's some "try" in the source code below which wants to return a potential error.

    const END: usize = 62501;  // 62501 for exactly 1M binary digits; usize is same like the Rust solution
    // const END: usize = 12;  // for testing

    // const M1: usize = 1_000_000;  // 1_000_000 / 62500 = 16
    // const M1: usize = (END-1) * 16;
    // const K250: usize = 250_000;
    // const K250: usize = (END-1) * 4;

    const m: usize = 65521;  // = 2^16 - 15
    const a: usize = 17364;
    const c: usize = 0;

    const file_bits_x   = "random_bitstring.bin";
    const file_bits_hex = "random_bitstring.byte";


    var x: [END]usize = undefined;  // allocate memory
    // also needed for the password


    // OK:
    // const seed = blk: {
    //     var seed: u64 = undefined;
    //     try std.posix.getrandom(std.mem.asBytes(&seed));
    //     break :blk seed;
    // };
    // var prng = std.Random.DefaultPrng.init(seed);  // OK!!
    // https://matklad.github.io/2025/03/19/comptime-zig-orm.html
    // https://zig.guide/master/standard-library/random-numbers/
    // x[0] = prng.random().uintLessThan(usize, m);  // OK!!
    // std.debug.print("x[0] = {d}\n", .{x[0]});  // for testing

    // so, why not using std.posix.getrandom(std.mem.asBytes(&seed)) directly as a random seed for x[0]? =>
    const seed = blk: {
        var seed: u64 = undefined;
        try std.posix.getrandom(std.mem.asBytes(&seed));
        break :blk seed;
    };

    // bringing seed/u64 into usize with maximum value m: type casting
    //   https://zig.guide/language-basics/floats/
    // 2025-12-19:
    const seed2 = @as(f32, @floatFromInt(seed)) / @as(f32, @floatFromInt(std.math.maxInt(u64))) * @as(f32, @floatFromInt(m-1)) + @as(f32, @floatFromInt(1));

    x[0] = @intFromFloat(seed2);
    // std.debug.print("x[0] = {d}\n", .{x[0]});  // for testing

    var buf16: [16]u8 = undefined;  // for several use cases in this program to hold a "string" of 16 "bits"
    var buf99: [99]u8 = undefined;  // for holding a "string" of 99 characters as user input from the keyboard until newline


    const allocator = std.heap.page_allocator;  // for several use cases in this program, not just here

    var bits_x:   std.ArrayList(u8) = .empty;  // 2025-12-19
    defer bits_x.deinit(allocator);            // 2025-12-19
    var bits_hex: std.ArrayList(u8) = .empty;  // 2025-12-19
    defer bits_hex.deinit(allocator);          // 2025-12-19


    std.debug.print("\ngenerating a random bit stream...\n", .{});
    var i: usize = 1;
    while (i < END) : (i += 1) {
        x[i] = (a*x[i-1] + c) % m;
        // std.debug.print("  {d:5}\n", .{x[i]});  // for testing; formatted printing
        // https://zig.guide/standard-library/advanced-formatting/

        const bits_x_str = try std.fmt.bufPrint(&buf16, "{b:0>16}", .{x[i]});  // with padding: 0001100110101101 --> 16 bits; type: []u8
        // https://github.com/ziglang/zig/issues/20078
        // std.debug.print("@typeName(bits_x_str) = {s}\n", .{@typeName(@TypeOf(bits_x_str))}); // for testing
        // https://ziggit.dev/t/how-typeof-chooses-the-resulting-type-when-there-are-more-than-one-arguments/3265
        // std.debug.print("bits_x_str = {s}\n", .{bits_x_str});  // for testing

        try bits_x.appendSlice(allocator, bits_x_str);  // 2025-12-19


        const bits_hex_str = try std.fmt.bufPrint(&buf16, "{x:0>4}", .{x[i]});  // with padding: a544 --> 2 bytes; type: []u8
        // std.debug.print("bits_hex_str = {s}\n", .{bits_hex_str});  // for testing

        try bits_hex.appendSlice(allocator, bits_hex_str);  // 2025-12-19

    }
    // std.debug.print("\nbits_x = {s}\n", .{bits_x.items});  // for testing
    // std.debug.print("bits_hex = {s}\n", .{bits_hex.items});  // for testing

    // for testing:
    // for (bits_x) |value| {
    //     std.debug.print("{any}", .{value});
    // }
    // std.debug.print("\n", .{});
    // for (bits_hex) |value| {
    //     std.debug.print("{any}", .{value});
    // }

    // write bit stream to disk:
    if (std.fs.cwd().createFile(file_bits_x, .{},)) |file| {
        defer file.close();
        if (file.writeAll( bits_x.items)) {
            std.debug.print("Bit stream has been written to disk under name:  {s}\n", .{file_bits_x});
        } else |err| {
            std.debug.print("could not write to file: {s} ! -- {any}\n", .{file_bits_x, err});
        }
    } else |err| {
        std.debug.print("could not create file: {s} ! -- {any}\n", .{file_bits_x, err});
    }
    // https://github.com/ziglang/zig/issues/5421#issuecomment-633720665

    // write byte stream to disk:
    if (std.fs.cwd().createFile(file_bits_hex, .{},)) |file| {
        defer file.close();
        if (file.writeAll( bits_hex.items)) {
            std.debug.print("Byte stream has been written to disk under name: {s}\n", .{file_bits_hex});
        } else |err| {
            std.debug.print("could not write to file: {s} ! -- {any}\n", .{file_bits_hex, err});
        }
    } else |err| {
        std.debug.print("could not create file: {s} ! -- {any}\n", .{file_bits_hex, err});
    }



    // make a password of N_CHAR printable chars: user input requested here
    var N_CHAR: i32 = 12;
    var answer: bool = false;
    // const stdin = std.io.getStdIn().reader();  // for version 0.14

    var stdin_reader = std.fs.File.stdin().reader(&buf99);  // 2025-12-19
    const stdin = &stdin_reader.interface;                  // 2025-12-19
    // see: https://stackoverflow.com/questions/62018241/current-way-to-get-user-input

    while (!answer) {
        N_CHAR = 12;
        const q1_str = try std.fmt.bufPrint(&buf16, "{}", .{N_CHAR});
        std.debug.print("\nPassword of {s} printable chars OK? 'y' or another integer number >= 8: ", .{q1_str});

        const answer_str = try stdin.takeDelimiterExclusive('\n');  // 2025-12-19
        // std.debug.print("\nanswer_str = {s}", .{answer_str});  // for testing
        stdin.tossBuffered();  // essential: "clear" this buffer!

        if ( std.mem.eql(u8, answer_str, "y")) {
            answer = true;
        } else {
            // const int_input = try std.fmt.parseInt(i32, answer_str, 10);  // this crashes on inputs like: 6.6, sadkjg, ...
            // https://github.com/ziglang/zig/blob/0.6.0/test/standalone/guess_number/main.zig
            const int_input = std.fmt.parseInt(i32, answer_str, 10) catch {
                std.debug.print("enter an integer number >= 8 or 'y'\n", .{});
                continue;  // this is the trick here!
            };

            // std.debug.print("\nnumber = {d}\n", .{int_input});  // for testing
            N_CHAR = int_input;
            if (N_CHAR < 8) {
                std.debug.print("enter an integer number >= 8 or 'y'\n", .{});
            } else {
                answer = true;
            }
        }
    }
    // std.debug.print("\nN_CHAR ={d}---\n", .{N_CHAR});  // for testing


    var WITH_SPECIAL_CHARS = true;
    answer = false;
    while (!answer) {
      std.debug.print("\nDo you want me to use special characters like .;,+*... ? 'y' or 'n': ", .{});

      const answer_str = try stdin.takeDelimiterExclusive('\n');  // 2025-12-19
      stdin.tossBuffered();  // essential: "clear" this buffer!

      // std.debug.print("\nanswer_str ={s}---\n", .{answer_str});  // for testing

      if ( std.mem.eql(u8, answer_str, "y")) {
          answer = true;
      } else {
          WITH_SPECIAL_CHARS = false;
          answer = true;
      }
    }


    // building char_set:
    //   https://zig.guide/standard-library/arraylist
    var char_set: std.ArrayList(u8) = .empty;  // 2025-12-19
    defer char_set.deinit(allocator);          // 2025-12-19


    var utf8_buffer = try allocator.alloc(u8, 1);  // allocate 1 byte for one UTF-8 character
    // https://character-encoding-decoding.mojoauth.com/utf-32-encoding--zig/

    if (WITH_SPECIAL_CHARS) {
      var codepoint: usize = 33;
      while (codepoint <= 127) : (codepoint += 1) {
        utf8_buffer[0] = @intCast(codepoint);

        try char_set.appendSlice(allocator, utf8_buffer);
      }
    } else {
      try char_set.appendSlice(allocator, "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz");
    }
    // std.debug.print("char_set = {s}\n", .{char_set.items});  // for testing
    // https://stackoverflow.com/questions/77290888/can-i-sprintf-to-an-arraylist-in-zig


    i = 0;             // char counter for the password
    var j: usize = 0;  // char counter for x

    var pw_chars: std.ArrayList(u8) = .empty;  // 2025-12-19
    defer pw_chars.deinit(allocator);          // 2025-12-19

    var char0_buffer_pw = try allocator.alloc(u8, 1);  // allocate 1 byte for 1 UTF-8 character
    var char1_buffer_pw = try allocator.alloc(u8, 1);
    const char_set1 = try char_set.toOwnedSlice(allocator);  // char_set is of type std.ArrayList(u8) --> convert to: []const u8
    // https://ziggit.dev/t/how-to-use-toownedslice-and-alternatives/2465

    while (i < N_CHAR) {
      const bin0 = try std.fmt.bufPrint(&buf16, "{b:0>16}", .{x[j]});  // with padding: 0001100110101101 --> 16 bits; type: []u8
      // std.debug.print("\nbin0 = {s}", .{bin0});  // for testing

      const bin0_0 = try std.fmt.parseInt(u8, bin0[0..8], 2);
      const bin0_1 = try std.fmt.parseInt(u8, bin0[8..16], 2);
      // https://ziggit.dev/t/how-to-convert-numerical-values-to-string-and-vice-versa-from-redit/719/2
      // std.debug.print("\nbin0_0 = {any}, bin0_1 = {any}", .{bin0_0, bin0_1});  // for testing: "bin0_0 = 180, bin0_1 = 5"

      char0_buffer_pw[0] = @as(u8, bin0_0);
      const char0_test = std.mem.containsAtLeast(u8, char_set1, 1, char0_buffer_pw);
      char1_buffer_pw[0] = @as(u8, bin0_1);
      const char1_test = std.mem.containsAtLeast(u8, char_set1, 1, char1_buffer_pw);
      // std.debug.print("\nchar0_test = {any}, char1_test = {any}\n", .{char0_test, char1_test});  // for testing

      if (char0_test) {
        try pw_chars.appendSlice(allocator, char0_buffer_pw);
        i += 1;
      }

      if (char1_test and i < N_CHAR) {
        try pw_chars.appendSlice(allocator, char1_buffer_pw);
        i += 1;
      }

      j += 1;
    }

    std.debug.print("\nYour password of {d} characters is: {s}\n", .{N_CHAR, pw_chars.items});
}

// end of random_bitstring_and_flexible_password_generator.zig

File: 03 - source code/01 - imperative languages/Lua/random_bitstring_and_flexible_password_generator.lua
--[[ random_bitstring_and_flexible_password_generator.lua

2025-07-14, 2025-07-23: fixing the string matching when creating the password; 2025-12-13: see below

run in Ubuntu 24 LTS:  $ lua random_bitstring_and_flexible_password_generator.lua


$ lua -v
Lua 5.4.8  Copyright (C) 1994-2025 Lua.org, PUC-Rio
$

]]


local math = require("math")
-- To optimize memory usage, prefer local variables over global ones.
-- Local variables are stored in the stack, which is faster and automatically cleaned up when no longer needed.
-- This practice minimizes the need for garbage collection.
-- https://dev.to/cristianalex_17/how-to-manage-memory-in-lua-in-2025-2a5b


------------------------------------------------------------------------
--
-- user defined functions

local STR_LENGTH_BIN = 16

local function Integer_to_bin_string(N)
    local bin_str = ""
    local j = STR_LENGTH_BIN
    local k = N
    while k > 0 and j >= 1 do
        if k % 2 == 0 then
            bin_str = '0'..bin_str
        else
            bin_str = '1'..bin_str
        end
        k = math.floor(k / 2)
        j = j - 1
    end
    bin_str = ('0'):rep(16-#bin_str)..bin_str
    return bin_str
end


local function safeStringToInt(str)
  return pcall(function()
    local num = tonumber(str)
    if not num then
      error("enter an integer number >= 8 or 'y'")
    elseif num and math.floor(num) == num then
      return num
    else
      error("enter an integer number >= 8 or 'y'")
    end
  end)
end

-- end of user defined functions
--
------------------------------------------------------------------------


local END = 62500  -- 62500 for exactly 1M binary digits
-- local END = 10  -- for testing

local m = 65521  -- = 2^16 - 15
local a = 17364
local c = 0

local file_bits_x   = 'random_bitstring.bin'
local file_bits_hex = 'random_bitstring.byte'

local x = {}  -- also needed for the password later

math.randomseed(os.time())
x[1] = math.random(1, m)  -- exclusive of m; 2025-12-13: (0, m) --> (1, m)

local bits_x   = {}  -- array of strings for the bit stream:  "0"'s + "1"'s
local bits_hex = {}  -- array of strings for the byte stream: [0..9][a..f]'s


print("\ngenerating a random bit stream...")

for i = 1, END do
  x[i+1] = (a * x[i] + c) % m
  -- print("\nx[i] = " .. x[i+1])  -- for testing

  local bits_x_str = Integer_to_bin_string(x[i+1])
  -- print("bits_x_str = " .. bits_x_str)  -- for testing
  bits_x[i] = bits_x_str

  local bits_hex_str = string.format("%04x", x[i+1])
  -- print("bits_hex_str = "..bits_hex_str)  -- for testing
  bits_hex[i] = bits_hex_str
end

local bits_x_str_total   = table.concat(bits_x, "")
local bits_hex_str_total = table.concat(bits_hex, "")
-- print("\nbits_x_str_total = "..bits_x_str_total)  -- for testing
-- print("\nbits_hex_str_total = "..bits_hex_str_total)  -- for testing

-- write bit stream to disk:
local file1 = io.open(file_bits_x, "w")
if file1 then
    file1:write(bits_x_str_total)
    file1:close()
    print("Bit stream has been written to disk under name:  "..file_bits_x)
else
    print("could not write to file: "..file_bits_x)
end

-- write byte stream to disk:
local file2 = io.open(file_bits_hex, "w")
if file2 then
    file2:write(bits_hex_str_total)
    file2:close()
    print("Byte stream has been written to disk under name: "..file_bits_hex)
else
    print("could not write to file: "..file_bits_hex)
end


-- make a password of n_char printable chars: user input requested here
local n_char = 12
local answer = false
local answer_str = ""

while answer == false do
  io.write("\nPassword of "..n_char.." printable chars OK? 'y' or another integer number >= 8: ")
  answer_str = io.read("*l")  -- read a line

  if answer_str == "y" then
    answer = true
  else
    local success, n_char_ = safeStringToInt(answer_str)
    if success then
      if n_char_ < 8 then
        print("enter an integer number >= 8 or 'y'")
      else
        n_char = n_char_
        answer = true
      end
    else
      print("enter an integer number >= 8 or 'y'")
   end
  end
end
-- print("n_char = "..n_char)  -- for testing


local with_special_chars = true
answer = false

while answer == false do
  io.write("\nDo you want me to use special characters like .;,+*... ? 'y' or 'n': ")
  answer_str = io.read("*l")  -- read a line

  if answer_str == "y" then
    answer = true
  else
    with_special_chars = false
    answer = true
  end
end
-- print("with_special_chars = "..tostring(with_special_chars))  -- for testing


local char_set = ""
if with_special_chars then
  for i = 33, 127 do  -- codepoints
    char_set = char_set .. string.char(i)
  end
else
  char_set = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
end
-- print ("char_set = "..char_set)  -- for testing


local i = 0
local j = 1
local pw_chars = ""
local bin0, bin0_0, bin0_1  = "", "", ""
local char0, char1 = "", ""

while i < n_char do
  bin0 = Integer_to_bin_string(x[j])
  -- print("\nbin0 = "..bin0)  -- for testing

  bin0_0 = string.sub(bin0, 1, 8)
  bin0_1 = string.sub(bin0, 9, 16)
  -- print("  bin0_0 = "..bin0_0.." -- bin0_1 = "..bin0_1)  -- for testing

  char0 = string.char(tonumber(bin0_0, 2))
  char1 = string.char(tonumber(bin0_1, 2))
  -- print("  char0 = "..char0.." -- char1 = "..char1)

  if string.match(char_set, "[%"..char0.."]") then  -- match also non-printable chars literally!
    pw_chars = pw_chars..char0
    i = i + 1
  end

  if string.match(char_set, "[%"..char1.."]") and string.len(pw_chars) < n_char then
    pw_chars = pw_chars..char1
    i = i + 1
  end

  j = j + 1
end

print ("\nYour password of "..n_char.." characters is: "..pw_chars)

-- end of random_bitstring_and_flexible_password_generator.lua

